File: src/main/java/com/letgo/shared/infrastructure/event/publisher/InMemorySyncDomainEventPublisher.java
package com.letgo.shared.infrastructure.event.publisher;

import com.letgo.shared.application.event.DomainEventPublisher;
import com.letgo.shared.application.event.DomainEventSubscriber;
import com.letgo.shared.domain.DomainEvent;

import java.util.List;

final public class InMemorySyncDomainEventPublisher implements DomainEventPublisher {
    private final List<DomainEventSubscriber> subscribers;

    public InMemorySyncDomainEventPublisher(List<DomainEventSubscriber> subscribers) {
        this.subscribers = subscribers;
    }

    @Override
    public void publish(List<DomainEvent> events) {
        events.forEach(event -> subscribers.forEach(subscriber -> tryToConsumeEvent(event, subscriber)));
    }

    private void tryToConsumeEvent(DomainEvent event, DomainEventSubscriber subscriber) {
        if (subscriber.isSubscribedTo(event)) {
            subscriber.consume(event);
        }
    }
}

File: src/main/java/com/letgo/shared/infrastructure/event/publisher/InMemorySyncDomainEventPublisher.kt
package com.letgo.shared.infrastructure.event.publisher

import com.letgo.shared.application.event.DomainEventPublisher
import com.letgo.shared.application.event.DomainEventSubscriber
import com.letgo.shared.domain.DomainEvent
import java.util.function.Consumer

class InMemorySyncDomainEventPublisher(
    private val subscribers: List<DomainEventSubscriber>
) : DomainEventPublisher {
    override fun publish(events: List<DomainEvent>) {
        events.forEach(Consumer { event: DomainEvent ->
            subscribers.forEach(Consumer { subscriber: DomainEventSubscriber ->
                tryToConsumeEvent(event, subscriber)
            })
        })
    }

    private fun tryToConsumeEvent(event: DomainEvent, subscriber: DomainEventSubscriber) {
        if (subscriber.isSubscribedTo(event)) {
            try {
                subscriber.consume(event)
            } catch (throwable: Throwable) {
                throwable.printStackTrace()
            }
        }
    }
}
